### ISSUE TITLE
DevTest: TGA test node textures are inconsistent

### ISSUE THREAD

**Issue by Wuzzy2** [minetest/minetest#12840]
##### Minetest version
```
5.6.1
```

##### Summary
The test textures for the TGA test nodes seem to have a couple of issues, making the test for TGA textures kind of broken.

First of all, one of the TGA test nodes in DevTest appears a flipped texture: `testnodes:tga_type2_16bpp_tb`. The texture appears to be mirrored vertically. All of the other test nodes have a non-flipped texture.

Second of all, there appears to be an inconsistent use of semitransparency among the test nodes. Some use alpha, some don't, and I don't see any system or logic behind that.

Third of all, there is TODO in the code:

```
TODO: Types 1, 2 & 10 should have two test nodes each (i.e. bottom-top
and top-bottom) for 16bpp (A1R5G5B5), 24bpp (B8G8R8), 32bpp (B8G8R8A8)
colors.
```

This is true, because some of the higher bpp variants are still missing.

##### What to do?

First of all, the flipped texture seems to be a problem in the texture itself, not with Minetest or IrrlichtMt. A simple edit in GIMP might fix this (however, I'm not sure if this is still a valid image that is exactly what we want to test for).

Second of all, probably this inconsistency could be solved by adding alpha and non-alpha variants more consistently.

Third of all, this can be obivously solved by adding the missing images but I don't know how to do that with the GIMP as the TGA test images are required to be in a very specific format. Probably I need some ImageMagick?

##### Steps to reproduce

* Place the various TGA test nodes in DevTest
* Look at them
* Also: Open the images in an image viewer

###### Appendix

Output of the `file` command on every TGA texture file:

```
testnodes_tga_type1_24bpp_bt.tga:  Targa image data - Map (4) 8 x 8 x 8
testnodes_tga_type1_24bpp_tb.tga:  Targa image data - Map (4) 8 x 8 x 8 +8 - top
testnodes_tga_type2_16bpp_bt.tga:  Targa image data - RGBA 8 x 8 x 16 - 1-bit alpha
testnodes_tga_type2_16bpp_tb.tga:  Targa image data - RGBA 8 x 8 x 16 - 1-bit alpha - top
testnodes_tga_type2_32bpp_bt.tga:  Targa image data - RGBA 8 x 8 x 32 - 8-bit alpha
testnodes_tga_type2_32bpp_tb.tga:  Targa image data - RGBA 8 x 8 x 32 +8 - 8-bit alpha - top
testnodes_tga_type3_16bpp_bt.tga:  Targa image data - Mono 8 x 8 x 16 - 8-bit alpha
testnodes_tga_type3_16bpp_tb.tga:  Targa image data - Mono 8 x 8 x 16 +8 - 8-bit alpha - top
testnodes_tga_type10_32bpp_bt.tga: Targa image data - RGBA - RLE 8 x 8 x 32 - 8-bit alpha
testnodes_tga_type10_32bpp_tb.tga: Targa image data - RGBA - RLE 8 x 8 x 32 +8 - 8-bit alpha - top
```

(I don't know if this is correct.)

**Comment by sfan5** [comment]
> First of all, one of the TGA test nodes in DevTest appears a flipped texture: `testnodes:tga_type2_16bpp_tb`. The texture appears to be mirrored vertically. All of the other test nodes have a non-flipped texture.

#12570

**Comment by Wuzzy2** [comment]
Oops. OK, I have removed this duplicate part from the issue and added a link.

**Comment by Zughy** [comment]
What am I supposed to see here, flipped node aside? Not gonna lie, I'm pretty confused by this .tga format, it's the first time I hear about it
![immagine](https://user-images.githubusercontent.com/63455151/203569018-97b3d85e-3e7f-4394-be24-3ba10b5e85b5.png)

**Comment by Wuzzy2** [comment]
Yes, that’s exaclty why I made this issue. It is unclear what each one of these is supposed to look like, so I have no idea whether the test rendering is correct.

Maybe all these nodes and textures should be thrown away and done from scratch, but this time with a clearly defined and documented "correct rendering".

I think a first step would be creating a full list of which TGA features should be tested, and then work from that. I don't really know how much of the TGA file format Minetest is meant to support in the first place. Anything intentionally unsupported should get a mention into `lua_api.txt` as well.

**Comment by erlehmann** [comment]
> Yes, that’s exaclty why I made this issue. It is unclear what each one of these is supposed to look like, so I have no idea whether the test rendering is correct.

This is kind of a chicken-egg problem, since TGA is usually used as the reference format in graphics libraries to make sure other stuff is correct. So TGA encoders are rarely wrong – but I remember that there was an issue with Irrlicht ignoring an alpha channel on some node.

> First of all, there appears to be an inconsistent use of semitransparency among the test nodes. Some use alpha, some don't, and I don't see any system or logic behind that.

There was no real system, I just created some nodes to try to cover as many code paths in as few nodes as possible.

> Maybe all these nodes and textures should be thrown away and done from scratch, but this time with a clearly defined and documented "correct rendering".

I think they`test nodes probably should be re-rendered using the `tga_encoder` mod. I'll see what I can do.

**Comment by erlehmann** [comment]
> > Maybe all these nodes and textures should be thrown away and done from scratch, but this time with a clearly defined and documented "correct rendering".
> 
> I think the test nodes probably should be re-rendered using the `tga_encoder` mod. I'll see what I can do.

You can see the start of my efforts to re-render the TGA test textures using the `tga_encoder` mod here:
https://git.minetest.land/erlehmann/tga_encoder/commit/e87a8fc332c7f33a3b6bd7fc24370284182c29a3

The ”correct rendering” I am aiming for so far is:

* 100% red letter T
* 100% green letter G
* 100% blue letter A
* background
** 50% transparent black (alpha: 128) for 32bpp images
** 100% transparent black (alpha 0) for 16bpp images
** black for 24bpp and 8bpp (no alpha channel)

Any better ideas?

**Comment by erlehmann** [comment]
@Wuzzy2  you can look at the current state of my script to generate several TGA test textures here:
https://git.minetest.land/erlehmann/tga_encoder/src/branch/master/tools/generate_test_textures.lua

Do you have opinions on this process?

**Comment by erlehmann** [comment]
Progress update: My test texture differ from the existing images byte-wise, so I still need to do some work.
But `testnodes_tga_type2_32bpp_tb.tga` and the generated `type2_32bpp_tb.tga ` differ only by one byte!

I am confident that I can figure the missing piece out.

https://git.minetest.land/erlehmann/tga_encoder/src/commit/aad231f5e406a7c1eaafe6595a64b9c55f230692/tools/generate_test_textures.lua

**Comment by Wuzzy2** [comment]
I like this script a lot. Apparently with GIMP (and possibly other image editors, too) you can't set much of the low-level stuff in TGA files, right? And it seems even with ImageMagick there are problems.
So yeah, it's probably best to generate them by script to be extra safe.

But please remember, if the textures still show up incorrectly in Minetest, this could either mean *your* textures are wrong *or* that Minetest has a rendering bug. So I suggest also testing these TGA files with external programs.

**Comment by erlehmann** [comment]
@Wuzzy2 I have refactored the script to use loops to generate all possible variants that `tga_encoder` can encode right now. Do you still like it? You can see the current state here: https://git.minetest.land/erlehmann/tga_encoder/pulls/1/files

> But please remember, if the textures still show up incorrectly in Minetest, this could either mean your textures are wrong or that Minetest has a rendering bug. So I suggest also testing these TGA files with external programs.

I have fixed one bug in the encoder already – A1R5G5B5 textures were not getting their alpha channel set correctly by `tga_encoder`.

I am testing all the test textures currently with sxiv, mtpaint and GIMP. It was quite interesting to see how some combinations of image properties just seem to break, even if the feature is otherwise not broken.

See the attached image for an example of GIMP tripping over itself. So far I am unsure why this happens, as both sxiv and mtpaint do not mangle the image (they do, however, show it without the alpha channel, which *might* indicate that the mistake is on my side):
![gimp-tga](https://github.com/minetest/minetest/assets/60905/f93a4dbc-8bc1-4624-85b7-ba572aa54de2)

I think that scenarios like this might also be the case for other (more complex) image formats and plan to investigate these kinds of parser differentials a bit more in the future. In particular, I consider it important to know which features of PNG Minetest does not support besides proper handling of the gamma information.
